pkg/lwip: Allow initializing different types of netifs

yarrick opened the pull request:
<!--
The RIOT community cares a lot about code quality.
Therefore, before describing what your contribution is about, we would like
you to make sure that your modifications are compliant with the RIOT
coding conventions, see https://github.com/RIOT-OS/RIOT/wiki/Coding-conventions.
-->

### Contribution description

<!--
Put here the description of your contribution:
- describe which part(s) of RIOT is (are) involved
- if it's a bug fix, describe the bug that it solves and how it is solved
- you can also give more information to reviewers about how to test your changes
-->

Split lwip netif setup into auto_init files, same as gnrc does.

To simplify init.c XFA can be used to group the init functions,
and then call them in a for loop. 

### Testing procedure

<!--
Details steps to test your contribution:
- which test/example to compile for which board and is there a 'test' command
- how to know that it was not working/available in master
- the expected success test output
-->

* `tests/lwip` still passes
*  Flash apps using lwIP (such as `examples/paho-mqtt`) and see that the expected interfaces show up in `ifconfig` and work

### Issues/PRs references

<!--
Examples: Fixes #1234. See also #5678. Depends on PR #9876.

Please use keywords (e.g., fixes, resolve) with the links to the issues you
resolved, this way they will be automatically closed when your pull request
is merged. See https://help.github.com/articles/closing-issues-using-keywords/.
-->

Depends on and includes #16153, same goal as #14382

yarrick commented:
Rebased to adapt to netdev cast removal

miri64 commented:
I like the general idea. However, to be true to the title, the IEEE 802.15.4 netifs should also get their own grouping.

yarrick commented:
Pushed new code finishing the move - removing RFC from title. 6LoWPAN devices are now also handled. I used blame to get copyrights for the different devices.

37 ifdefs died in the process.

benpicco commented:
This needs another rebase

benpicco commented:
#### sam0_eth

```
2021-08-19 12:00:57,577 # ifconfig
2021-08-19 12:00:57,582 # Iface ET0 HWaddr: a6:ed:29:ec:cb:f3 Link: up State: up
2021-08-19 12:00:57,584 #         Link type: wired
2021-08-19 12:00:57,590 #         inet addr: 192.168.42.112 mask: 255.255.255.0 gw: 192.168.42.1
2021-08-19 12:00:57,596 #         inet6 addr: fe80:0:0:0:a4ed:29ff:feec:cbf3 scope: link
2021-08-19 12:00:57,602 #         inet6 addr: 2001:16b8:457d:f100:a4ed:29ff:feec:cbf3 scope: global
```

#### nrf802154

```
2021-08-19 12:02:23,999 # ifconfig
2021-08-19 12:02:24,005 # Iface L60 HWaddr: aa:cc:f4:7b:67:c4:ba:f1 Link: up State: up
2021-08-19 12:02:24,007 #         Link type: wireless
2021-08-19 12:02:24,013 #         inet6 addr: fe80:0:0:0:a8cc:f47b:67c4:baf1 scope: link
```

#### at86rf2xx

```
2021-08-19 12:05:19,378 # ifconfig
2021-08-19 12:05:19,384 # Iface L60 HWaddr: 00:04:25:19:18:01:c9:05 Link: up State: up
2021-08-19 12:05:19,387 #         Link type: wireless
2021-08-19 12:05:19,392 #         inet6 addr: fe80:0:0:0:204:2519:1801:c905 scope: link
```

benpicco commented on pkg/lwip/include/lwip.h:
```diff
@@ -31,6 +33,11 @@ extern "C" {
 
+struct netif;
+
+struct netif *lwip_add_ethernet(struct netif *netif, netdev_t *state);
+struct netif *lwip_add_6lowpan(struct netif *netif, netdev_t *state);
```
I'm surprised doccheck doesn't complain about the missing documentation :wink:

benpicco commented on pkg/lwip/init_devs/init.c:
```diff
@@ -0,0 +1,88 @@
+    /* Ethernet interfaces
+     * ------------------- */
+
+    if (IS_USED(MODULE_ATWINC15x0)) {
```
```suggestion
    if (IS_USED(MODULE_ATWINC15X0)) {
```

yarrick replied:
Thanks, fixed

benpicco commented:
Please add doxygen for `lwip_add_ethernet()` and `lwip_add_6lowpan()` - you can squash directly

yarrick commented:
OK, will do

yarrick commented:
Moved `lwip_add_ethernet()` and `lwip_add_6lowpan()` to init.c and added a new headerfile for them.

Added XFA for finding the available init functions, it seems to work.

benpicco reviewed (approved):
Nice cleanup and great to have some users for XFA too!
